### PR TITLE
[Feature] Add Mosaic Data Augmentation

### PR DESCRIPTION
**Pull request by Adamdad**
- [x] Add mosaic data augmentation that combine 4 sub-image into 1 full image
- [x] Reformat the code
- [x] Image visualization
- [x] Add unit-test
- [x] Add config file for YOLOv3
- [ ] Benchmark on COCO and VOC with/without mosaic augmentation

### CODE REVIEW

File: mmdet/datasets/pipelines/transforms.py [hunk 1]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
                      else f'cutout_shape={self.candidates}, ')
         repr_str += f'fill_in={self.fill_in})'
         return repr_str
+
+@PIPELINES.register_module()
+class Mosaic(object):
+    """Mosaic augmentation.
+    (0:cut_y, 0:cut_x)                        -- image 0
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1816]: Please refer to RandomCenterCropPad for a more illustrative description.

File: mmdet/datasets/pipelines/transforms.py [hunk 2]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+    (0:cut_y, cut_x:self.size[1])             -- image 1
+    (cut_y:self.size[0], 0:cut_x)             -- image 2
+    (cut_y:self.size[0], cut_x:self.size[1])  -- image 3
+    The output image is composed of the parts from each sub-image.
+    Args:
+        size (tuple): Output size of the image. Default: (640, 640).
+        jitter (float): It decides the size of the cropping window.
+        min_offset (float): Volume of the offset of the cropping window.
+        load_pipeline (list): Pipeline for loading images and annotations.
+        letter_box(bool): Whether to maintain the aspect ratio of the
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1826]: Use `keep_aspect_ratio` explicitly.

File: mmdet/datasets/pipelines/transforms.py [hunk 3]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+        out_labels = []
+        out_ignores = []
+
+        for i in range(4):
+            if i == 0:
+                # use the current image
+                results_i = results
+            else:
+                # randomly sample a new image from the dataset
+                results_i = next(self.dataset)
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1895]: Better use `np.random.randint(0, len(self.dataset), size=3)` before the `for` loop.

File: mmdet/datasets/pipelines/transforms.py [hunk 4]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+            int(self.size[1] * (1 - self.min_offset[1]))
+        )
+
+        buffer_img = np.zeros((self.size[0], self.size[1], 3), dtype=np.float32)
+
+        out_bboxes = []
+        out_labels = []
+        out_ignores = []
+
+        for i in range(4):
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1889]: We may use `for i in ('top_left', 'top_right', 'bottom_left', bottom_right')` to explicitly express its position.

File: mmdet/datasets/pipelines/transforms.py [hunk 5]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+                return None
+            results[key] = bboxes[valid_inds, :]
+            # label fields. e.g. gt_labels and gt_labels_ignore
+            label_key = self.bbox2label.get(key)
+            if label_key in results:
+                results[label_key] = results[label_key][valid_inds]
+
+        return results
+
+    def _filter_truth(self, bboxes, labels, crop_w, crop_h, mix_x, mix_y):
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:2048]: Rename it to `adjust_coordinate`

File: mmdet/datasets/pipelines/transforms.py [hunk 6]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+                                           crop_w=self.size[1] - cut_x,
+                                           crop_h=self.size[0] - cut_y,
+                                           mix_x=cut_x,
+                                           mix_y=cut_y)
+            tmp_img[cut_y:, cut_x:] = results_i['img'].copy()
+
+        return tmp_img, bboxes, labels, ignores
+
+    def _crop_data(self, results, crop_size, allow_negative_crop):
+        """Function to randomly crop images, bounding boxes, masks, semantic
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1995]: We may maintain a `RandomCrop` object and add a `set_crop_size` method in this class. Therefore, this `_crop_data` method is not necessary any more.

File: mmdet/datasets/pipelines/transforms.py [hunk 7]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
                      else f'cutout_shape={self.candidates}, ')
         repr_str += f'fill_in={self.fill_in})'
         return repr_str
+
+@PIPELINES.register_module()
+class Mosaic(object):
+    """Mosaic augmentation.
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1815]: Docstring style:
```python
"""Main description.

Detailed description
Newline
Newline. 

Args:
"""

```

File: mmdet/datasets/pipelines/transforms.py [hunk 8]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+    (0:cut_y, 0:cut_x)                        -- image 0
+    (0:cut_y, cut_x:self.size[1])             -- image 1
+    (cut_y:self.size[0], 0:cut_x)             -- image 2
+    (cut_y:self.size[0], cut_x:self.size[1])  -- image 3
+    The output image is composed of the parts from each sub-image.
+    Args:
+        size (tuple): Output size of the image. Default: (640, 640).
+        jitter (float): It decides the size of the cropping window.
+        min_offset (float): Volume of the offset of the cropping window.
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1824]: min_offset (float | tuple[float] | None)

Comment (Johnson-Wang) [reply mmdet/datasets/pipelines/transforms.py:1824]: Default: 0.2.

File: mmdet/datasets/pipelines/transforms.py [hunk 9]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+    """
+
+    def __init__(self,
+                 size=(640, 640),
+                 min_offset=0.2,
+                 dataset=None,
+                 allow_negative_crop=False,
+                 bbox_clip_border=True):
+        assert isinstance(size, tuple)
+        assert size[0] > 0 and size[1] > 0, (
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1837]:
```
if size[0] <=0 or size[1] <= 0:
    raise ValueError('')
```

File: mmdet/datasets/pipelines/transforms.py [hunk 10]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+            'size must > 0 in train mode')
+        assert isinstance(size[0], int) and isinstance(
+            size[1], int)
+
+        if min_offset is None:
+            self.min_offset = (0, 0)
+        else:
+            if isinstance(min_offset, float):
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1845]: elif

File: mmdet/datasets/pipelines/transforms.py [hunk 11]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+        self.allow_negative_crop = allow_negative_crop
+        self.dataset = dataset
+        self.bbox_clip_border = bbox_clip_border
+        self.bbox2label = {
+            'gt_bboxes': 'gt_labels',
+            'gt_bboxes_ignore': 'gt_labels_ignore'
+        }
+
+    def __call__(self, results):
+        """Call the function to mix 4 images
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1866]: Blank line.

File: mmdet/datasets/pipelines/transforms.py [hunk 12]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+        results['gt_labels'] = out_labels
+        results['gt_bboxes_ignore'] = out_ignores
+        return results
+
+    def _mosiac_combine(self, i, results_i, tmp_img, cut_x, cut_y):
+        """
+        Args:
+            i (int): Index for the subimage, i = 0,1,2,3
+            results_i (dict): Result dict from loading pipeline.
+            tmp_img (numpy array, (H x W x 3)): buffer for mosiac image
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1918]: tmp_img -> img.

File: mmdet/datasets/pipelines/transforms.py [hunk 13]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+        results['gt_bboxes'] = out_bboxes
+        results['gt_labels'] = out_labels
+        results['gt_bboxes_ignore'] = out_ignores
+        return results
+
+    def _mosiac_combine(self, i, results_i, tmp_img, cut_x, cut_y):
+        """
+        Args:
+            i (int): Index for the subimage, i = 0,1,2,3
+            results_i (dict): Result dict from loading pipeline.
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1917]: results_i -> results.

File: mmdet/datasets/pipelines/transforms.py [hunk 14]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+            tmp_img (numpy array, (H x W x 3)): buffer for mosiac image
+            cut_x (int): dividing coordinate x
+            cut_y (int): dividing coordinate y
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1920]: Blank line.

File: mmdet/datasets/pipelines/transforms.py (hunk 5 above)
Comment (Johnson-Wang) [reply mmdet/datasets/pipelines/transforms.py:2048]: Use `results` as its input.

File: mmdet/datasets/pipelines/transforms.py (hunk 12 above)
Comment (Johnson-Wang) [reply mmdet/datasets/pipelines/transforms.py:1918]: `cut_x, cut_y` -> `cut_position`.

File: mmdet/datasets/pipelines/transforms.py [hunk 15]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+        bboxes[:, 0::2] = np.clip(bboxes[:, 0::2], 0, crop_w)
+        bboxes[:, 1::3] = np.clip(bboxes[:, 1::3], 0, crop_h)
+
+        # keep the reasonable sized bboxes and labels
+        reserved_inds = (bboxes[:, 2] > bboxes[:, 0]) & (
+            bboxes[:, 3] > bboxes[:, 1])
+
+        bboxes = bboxes[reserved_inds]
+        if labels is not None:
+            labels = labels[reserved_inds]
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:2074]: Not necessary.

File: mmdet/datasets/pipelines/transforms.py [hunk 16]
```diff
@@ -1809,3 +1809,271 @@ def __repr__(self):
...
+        """
+        Args:
+            i (int): Index for the subimage, i = 0,1,2,3
+            results_i (dict): Result dict from loading pipeline.
+            tmp_img (numpy array, (H x W x 3)): buffer for mosiac image
+            cut_x (int): dividing coordinate x
+            cut_y (int): dividing coordinate y
+        Returns:
+            bboxes: Result dict with images and bounding boxes cropped.
+        """
```
Comment (Johnson-Wang) [inline mmdet/datasets/pipelines/transforms.py:1923]: `cut_size = [(cut_y, cut_x), (cut_y, self.size[1] - cut_x), ...]  # annotation`

Comment (Johnson-Wang) [reply mmdet/datasets/pipelines/transforms.py:1923]: `slices = [(Slice(0, cut_y), Slice(0, cut_x)), ...] `

File: mmdet/datasets/pipelines/transforms.py (hunk 6 above)
Comment (Johnson-Wang) [reply mmdet/datasets/pipelines/transforms.py:1995]: Add support for segmentation.

**Comment by Johnson-Wang** [comment]
We may add a config for yolo-v3 as well.

File: mmdet/datasets/pipelines/transforms.py (hunk 16 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py:1955]: revised

File: mmdet/datasets/pipelines/transforms.py (hunk 15 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: deleted

File: mmdet/datasets/pipelines/transforms.py (hunk 14 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py (hunk 13 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py (hunk 12 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py (hunk 5 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py (hunk 4 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py (hunk 3 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py (hunk 1 above)
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: revised

**Review by Johnson-Wang — COMMENTED** [review]
Please also add a config for yolo if you could.

File: mmdet/datasets/pipelines/transforms.py [hunk 17]
```diff
@@ -1809,3 +1827,178 @@ def __repr__(self):
...
+        out_ignores = []
+
+        for loc in ('top_left', 'top_right', 'bottom_left', 'bottom_right'):
+            if loc == 'top_left':
+                # use the current image
+                results_i = copy.deepcopy(results)
+            else:
+                # randomly sample a new image from the dataset
+                index = random.randint(self.num_sample)
+                results_i = copy.deepcopy(self.dataset.__getitem__(index))
```
[the remark this answers is not in the document]
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py:1922]: pytorch dataset is not a iterator. Direct indexing does not work.

**Comment by Adamdad** [comment]
img

**Comment by Adamdad** [comment]
img

**Comment by leehao178** [comment]
It's great! Could you add example Config, i don't know how to add dataset
Thank you so much for your work!!

File: mmdet/datasets/pipelines/transforms.py [hunk 18]
```diff
@@ -705,14 +705,31 @@ def _crop_data(self, results, crop_size, allow_negative_crop):
...
 
         return results
 
+    def set_crop_size(self, crop_size):
+        """Reset the crop_size attribute.
+
+        Args:
+            crop_size (tuple): (h, w).
+
+        Returns:
```
[the remark this answers is not in the document]
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py [hunk 19]
```diff
@@ -1809,3 +1827,178 @@ def __repr__(self):
...
+                 min_offset=0.2,
+                 dataset=None):
+
+        assert isinstance(size, tuple)
+        assert isinstance(size[0], int) and isinstance(
+            size[1], int)
+        if size[0] <= 0 or size[1] <= 0:
+            raise ValueError('image size must > 0 in train mode')
+
+        if min_offset is None:
```
[the remark this answers is not in the document]
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py [hunk 20]
```diff
@@ -1809,3 +1827,178 @@ def __repr__(self):
...
+            self.min_offset = (0, 0)
+        elif isinstance(min_offset, float):
+            assert 0 <= min_offset <= 1
+            self.min_offset = (min_offset, min_offset)
+        elif isinstance(min_offset, tuple):
+            assert isinstance(min_offset[0], float) \
+                   and isinstance(min_offset[1], float)
```
[the remark this answers is not in the document]
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py:1855]: fixed

File: mmdet/datasets/pipelines/transforms.py [hunk 21]
```diff
@@ -1809,3 +1827,178 @@ def __repr__(self):
...
+            int(self.size[0] * self.min_offset[0]),
+            int(self.size[0] * (1 - self.min_offset[0]))
+        )
+        cut_x = random.randint(
+            int(self.size[1] * self.min_offset[1]),
+            int(self.size[1] * (1 - self.min_offset[1]))
+        )
+
+        cut_position = (cut_y, cut_x)
+        tmp_img = np.zeros((self.size[0], self.size[1], 3), dtype=np.float32)
```
[the remark this answers is not in the document]
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py [hunk 22]
```diff
@@ -1809,3 +1827,178 @@ def __repr__(self):
...
+
+        elif loc == 'bottom_right':
+            # Image 3: bottom right
+            crop_size = (self.size[0] - cut_position[0], self.size[1] - cut_position[1])
+            img_slices = (slice(cut_position[0], self.size[0]), slice(cut_position[1], self.size[1]))
+            paste_position = cut_position
+
+        self.cropper.set_crop_size(crop_size)
+        results = self.cropper(results)
+        img[img_slices] = results['img'].copy()
```
[the remark this answers is not in the document]
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py]: fixed

File: mmdet/datasets/pipelines/transforms.py [hunk 23]
```diff
@@ -1809,3 +1827,178 @@ def __repr__(self):
...
+         Args:
+            results (dict): Result dict from :obj:`dataset`.
+            paste_position (tuple[int]): paste up-left corner coordinate (y, x)
+                in moaiac image.
+
+        Returns:
+            results (dict): Result dict with corrected bbox and mask coordinate.
+        """
+
+        for key in results.get('bbox_fields', []):
```
[the remark this answers is not in the document]
Comment (Adamdad) [reply mmdet/datasets/pipelines/transforms.py:1991]: added

File: mmdet/datasets/pipelines/transforms.py [hunk 24]
```diff
@@ -1809,3 +1809,189 @@ def __repr__(self):
...
+
+        Returns:
+            results (dict): Result dict with corrected bbox
+                and mask coordinate.
+        """
+
+        for key in results.get('bbox_fields', []):
+            box = results[key]
+            box[:, 0::2] += paste_position[1]
+            box[:, 1::3] += paste_position[0]
```
Comment (HuangJunJie2017) [inline mmdet/datasets/pipelines/transforms.py:1994]: box[:, 1::2] += paste_position[0]

**Comment by Hyerin-oh** [comment]
Hello! I'm using a translator, so English may be unnatural. I ask of your understanding.
I'm sorry if it's not polite to give you a comment.
I happened to see this PR while passing by. Thank you for the nice code!
What do you think about adding a probablity? like this.

```
@PIPELINES.register_module()
class Mosaic(object):
    def __init__(self, size=(512, 512), min_offset=0.2, prob = 0.5 , dataset=None):
        assert isinstance(size, tuple)
        assert isinstance(size[0], int) and isinstance(size[1], int)
        assert 0 <= prob <= 1
        if size[0] <= 0 or size[1] <= 0:
            raise ValueError('image size must > 0 in train mode')
        if isinstance(min_offset, float):
            assert 0 <= min_offset <= 1
            self.min_offset = (min_offset, min_offset)
        elif isinstance(min_offset, tuple):
            assert isinstance(min_offset[0], float) \
                   and isinstance(min_offset[1], float)
            assert 0 <= min_offset[0] <= 1 and 0 <= min_offset[1] <= 1
            self.min_offset = min_offset
        else:
            raise TypeError('Unsupported type for min_offset, '
                            'should be either float or tuple[float]')
        self.size = size
        self.dataset = dataset
        self.cropper = RandomCrop(crop_size=size)
        self.num_sample = len(dataset)
        self.prob = prob

    def __call__(self,results):
        if random.uniform(0, 1) > self.prob:
            return results
```

I've been working on a few lines of code, so please consider it. Thank You

**Comment by likyoo** [comment]
Thanks for your code.
Could you provide a config file about how to use "Mosaic" ?

**Comment by Adamdad** [comment]
> Hello! I'm using a translator, so English may be unnatural. I ask of your understanding.
> I'm sorry if it's not polite to give you a comment.
> I happened to see this PR while passing by. Thank you for the nice code!
> What do you think about adding a probablity? like this.
> 
> ```
> @PIPELINES.register_module()
> class Mosaic(object):
>     def __init__(self, size=(512, 512), min_offset=0.2, prob = 0.5 , dataset=None):
>         assert isinstance(size, tuple)
>         assert isinstance(size[0], int) and isinstance(size[1], int)
>         assert 0 <= prob <= 1
>         if size[0] <= 0 or size[1] <= 0:
>             raise ValueError('image size must > 0 in train mode')
>         if isinstance(min_offset, float):
>             assert 0 <= min_offset <= 1
>             self.min_offset = (min_offset, min_offset)
>         elif isinstance(min_offset, tuple):
>             assert isinstance(min_offset[0], float) \
>                    and isinstance(min_offset[1], float)
>             assert 0 <= min_offset[0] <= 1 and 0 <= min_offset[1] <= 1
>             self.min_offset = min_offset
>         else:
>             raise TypeError('Unsupported type for min_offset, '
>                             'should be either float or tuple[float]')
>         self.size = size
>         self.dataset = dataset
>         self.cropper = RandomCrop(crop_size=size)
>         self.num_sample = len(dataset)
>         self.prob = prob
> 
>     def __call__(self,results):
>         if random.uniform(0, 1) > self.prob:
>             return results
> ```
> 
> I've been working on a few lines of code, so please consider it. Thank You

**Comment by Adamdad** [comment]
> Hello! I'm using a translator, so English may be unnatural. I ask of your understanding.
> I'm sorry if it's not polite to give you a comment.
> I happened to see this PR while passing by. Thank you for the nice code!
> What do you think about adding a probablity? like this.
> 
> ```
> @PIPELINES.register_module()
> class Mosaic(object):
>     def __init__(self, size=(512, 512), min_offset=0.2, prob = 0.5 , dataset=None):
>         assert isinstance(size, tuple)
>         assert isinstance(size[0], int) and isinstance(size[1], int)
>         assert 0 <= prob <= 1
>         if size[0] <= 0 or size[1] <= 0:
>             raise ValueError('image size must > 0 in train mode')
>         if isinstance(min_offset, float):
>             assert 0 <= min_offset <= 1
>             self.min_offset = (min_offset, min_offset)
>         elif isinstance(min_offset, tuple):
>             assert isinstance(min_offset[0], float) \
>                    and isinstance(min_offset[1], float)
>             assert 0 <= min_offset[0] <= 1 and 0 <= min_offset[1] <= 1
>             self.min_offset = min_offset
>         else:
>             raise TypeError('Unsupported type for min_offset, '
>                             'should be either float or tuple[float]')
>         self.size = size
>         self.dataset = dataset
>         self.cropper = RandomCrop(crop_size=size)
>         self.num_sample = len(dataset)
>         self.prob = prob
> 
>     def __call__(self,results):
>         if random.uniform(0, 1) > self.prob:
>             return results
> ```
> 
> I've been working on a few lines of code, so please consider it. Thank You

Will consider, thanks!

**Comment by likyoo** [comment]
Hello, I look forward to your follow-up work very much.
Could you provide an example of how to define "dataset" in the configuration file?
By the way, I think it would be better if it can support multi-scale training.

**Comment by ZwwWayne** [comment]
Mosaic augmentation has been supported in master branch with YOLOX

**Pull request closed without merging**